File: Ahorn/triggers/spawnGravityTrigger.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperSpawnGravityTrigger

using ..Ahorn, Maple

const PLUGIN_VERSION = "1"

@mapdef Trigger "GravityHelper/SpawnGravityTrigger" SpawnGravityTrigger(
    x::Integer, y::Integer,
    width::Integer=Maple.defaultTriggerWidth, height::Integer=Maple.defaultTriggerHeight,
    pluginVersion::String=PLUGIN_VERSION,
    gravityType::Integer=0, fireOnBubbleReturn::Bool=true
)

const gravityTypes = Dict{String, Integer}(
    "Normal" => 0,
    "Inverted" => 1,
)

const placements = Ahorn.PlacementDict(
    "Spawn Gravity Trigger (Normal) (GravityHelper)" => Ahorn.EntityPlacement(
        SpawnGravityTrigger,
        "rectangle",
        Dict{String, Any}(
            "gravityType" => 0,
        )
    ),
    "Spawn Gravity Trigger (Inverted) (GravityHelper)" => Ahorn.EntityPlacement(
        SpawnGravityTrigger,
        "rectangle",
        Dict{String, Any}(
            "gravityType" => 1,
        )
    ),
)

Ahorn.editingIgnored(trigger::SpawnGravityTrigger, multiple::Bool=false) = String["modVersion", "pluginVersion"]

Ahorn.editingOptions(trigger::SpawnGravityTrigger) = Dict{String, Any}(
    "gravityType" => gravityTypes
)

end

File: Loenn/triggers/spawnGravityTrigger.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")

local placementData = helpers.createPlacementData('1', {
    width = 8,
    height = 8,
    gravityType = consts.gravityTypes.normal.index,
    fireOnBubbleReturn = true,
})

local spawnGravityTrigger = {
    name = "GravityHelper/SpawnGravityTrigger",
    ignoredFields = consts.ignoredFields,
    fieldInformation = {
        gravityType = consts.fieldInformation.gravityType(0,1),
    },
    placements = {
        {
            name = "normal",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.normal.index,
            }),
        },
        {
            name = "inverted",
            data = helpers.union(placementData, {
                gravityType = consts.gravityTypes.inverted.index,
            }),
        },
    },
}

return spawnGravityTrigger
